### ISSUE TITLE
Update master from dev (nlog 5)

### ISSUE THREAD

**Issue by 304NotModified** [NLog/NLog#4931]
Master is behind currently. 

Maybe also remove dev after this, and change to master. (Needs some updates to the docs and branch rules/protection)

**Comment by snakefoot** [comment]
Really hate source-control-merges. What about making a release-cuts ?
- create nlog4-branch from master
- create nlog5-branch from dev

And then have "main" for any work related to nlog6

**Comment by 304NotModified** [comment]
We don't need those branches as we have tags?

Only if necessary hotfixes are expected for nlog 4, a branch would be nice.

**Comment by snakefoot** [comment]
I like that one only have to make a commit to a branch, and then automatically a new version is released.

**Comment by 304NotModified** [comment]
Pa. NLog 5 links to the changelog on master 

https://www.nuget.org/packages/NLog/5.0.0

**Comment by snakefoot** [comment]
I can see with all major version of NLog, then there have been a need to patch the previous major-version.

I think we should continue to develop in dev-branch, but when the time arrives where pull-requests start piling up, that introduces wanted breaking changes, then nlog5-branch should be created with focus on bug-fixes (maintenance-mode-only with cherry picking).

After the adventure with NLog v4 and v5, then I see that the Nlog-git-repository has become huge mess, because of multiple huge branch-merges between dev and master.

Right now using minor-builds for NLog v5 (Ex. v5.2) for marking unwanted API as obsolete (but not making breaking changes),  and guiding the NLog-users towards the future API of NLog v6.

**Comment by snakefoot** [comment]
Started to think about what NLog v6 should be, and made some goals:

- https://github.com/NLog/NLog.github.io/blob/ver6/_posts/2024-10-01-nlog-6-0-goals.md

NLog v5 has been out for 2 years, and only received a few complaints about the changes in default values. But probably just hearing about the tip of the iceberg. But 2 years is also a short time compared to NLog v4 which has been around for 7 years, while waiting for NLog v3 users to move over.

Seems people are not completely done with NLog v4, so probably need to continue the support for 1-2 years more. But at the same time would like to start working on the NLog v6 release. Will probably not have final build ready for the NET9 release, but at least have some kind of NLog v6 preview-build.

This means putting NLog v5 into maintenance mode (together with NLog v4).

**Comment by snakefoot** [comment]
Make first draft the NLog v6 release notes:

- https://github.com/NLog/NLog.github.io/blob/ver6_changes/_posts/2025-04-01-nlog-6-0-major-changes.md

**Comment by snakefoot** [comment]
NLog v6 includes changes, that will require [wiki-updates](https://github.com/NLog/NLog/pulls?q=is%3Apr+is%3Amerged+milestone%3A%226.0%22+label%3A%22needs+documentation+on+wiki%22+-label%3A%22documentation+done%22+)

**Comment by snakefoot** [comment]
> @topsterde Why isn't there a separate NLog.Config package? This would then contain the LoadConfigurationFromFile extension method and the XmlParser NLog.config logic.

Thank you for your input. I have tried to explain my view with this paragraph in the release-notes:

> If the `NLog.config`-file had to be explicitly loaded, then the AOT-build could trim much more, since right now the AOT-build cannot predict what types will be required by the NLog.config-file. But disabling automatic loading of the `NLog.config`-file is a huge breaking change, and would hurt lots of existing applications.

Again the goal is to make the upgrade path to NLog v6 as painless as possible for all existing consumers, but at the same time support AOT-builds without build-warnings for those who need. This puts some restrictions on how much can be carved away from the NLog-package without hurting the expected/existing behavior.

> Unforunately I can't find a nightly nuget feed to test V6. Is that available?

You can find nuget-package-preview here (Just download to local nuget-folder and change file-extension from `.zip` to `.nupkg`)

- https://ci.appveyor.com/project/nlog/nlog/builds/51965108/job/by4u9xpo6vyga6hr/artifacts

**Comment by snakefoot** [comment]
NLog v6.0 Preview1 has now been released, so people can try out the new packages:
- https://www.nuget.org/packages/NLog.Targets.AtomicFile
- https://www.nuget.org/packages/NLog.Targets.GZipFile
- https://www.nuget.org/packages/NLog.Targets.Mail
- https://www.nuget.org/packages/NLog.Targets.Network
- https://www.nuget.org/packages/NLog.Targets.Trace
- https://www.nuget.org/packages/NLog.Targets.WebService
- https://www.nuget.org/packages/NLog.Targets.RegEx

**Comment by snakefoot** [comment]
Not so much feedback so far for NLog v6, but happy that some can see improvements:

> **Update NLog to 6.0.0-preview1**
> I normally wouldn't update to a prerelease version, but this removes 10MB
from the size of an AOT-compiled executable (40MB -> 30MB) and eliminates
a bunch of warnings related to System.Linq.Expressions.
> https://github.com/Helion-Engine/Helion/commit/bf932967cbcedc9f2d4ad120ee67f471fb86416f

> Can confirm that with the 6.0.0 preview there are no more warnings
> https://github.com/NLog/NLog/issues/5258#issuecomment-2883836811

**Comment by topsterde** [comment]
My feedback... The migration was completed in 5 minutes and was very well documented. All the requested features worked very well. I think the progress is very good. I like that the NLog component has been reduced in size. I'm looking forward to the release.

**Comment by snakefoot** [comment]
Thank you for the confirmation. Also feel that NLog v6 is getting close to a release-candidate.

I have decided not to perform the cleanup of the NLog Logger API, but instead added `[OverloadResolutionPriority(-1)]` to all legacy methods. This should help all users upgrading to NET9 + NET10 to move away from the legacy methods, and instead use generic-methods or `ReadOnlySpan`. Then next major version of NLog can have the honors of doing the cleanup without too much pain.

Just completed adding `<Nullable>enable</Nullable>` to the NLog.csproj, but still need to apply the project-setting to the other nuget-package-libraries. And when that has been completed, then need to fix `NLog.Targets.WebService`-csproj, so it doesn't depend on `System.Xml`.

**Comment by snakefoot** [comment]
[NLog v6.0 RC1](https://github.com/NLog/NLog/releases/tag/v6.0.0-rc1) has now been released on [nuget](https://www.nuget.org/packages/NLog/6.0.0-rc1).

Not planning to bring any more breaking changes into the NLog v6 milestone.

Unless something bad is discovered, then this will be the last preview-release, before the official release.

**Comment by snakefoot** [comment]
[NLog v6.0 RC2](https://github.com/NLog/NLog/releases/tag/v6.0.0-rc2) has now been released on [nuget](https://www.nuget.org/packages/NLog/6.0.0-rc2).

Discovered a bug in the new XmlParser where it failed to handle XML-comments inside XML-processing-instructions.

Also changed NLog ConsoleTarget to not use WriteLine by default, but instead write directly to Console-output-buffer for improved performance (Twice as fast with `<targets async="true">` as it allows batching and reduces string-allocations).

Have now also released [NLog.Extensions.Logging v6.0 RC2](https://github.com/NLog/NLog.Extensions.Logging/releases/tag/v6.0.0-rc2) on [nuget](https://www.nuget.org/packages/NLog.Extensions.Logging/6.0.0-rc2#releasenotes-body-tab)

**Comment by Balkoth** [comment]
Can you please comment on this from the migration doc:

- ArchiveDateFormat - Marked as obsolete. Instead use new ArchiveSuffixFormat
- ArchiveNumbering - Marked as obsolete. Instead use new ArchiveSuffixFormat (Rolling is unsupported).

`ArchiveSuffixFormat` is nowhere documented and on v5 i have set it up like this:

```json
"FileTarget": {
  "type": "File",
  "fileName": "C:\\Logfiles\\AppLog.txt",
  "archiveFileName": "C:\\Logfiles\\AppLog_{#}.txt",
  "archiveNumbering": "Date",
  "archiveEvery": "Day",
  "archiveDateFormat": "yyyyMMdd"
}
```

How do i have to setup `ArchiveSuffixFormat` to keep it that way? I really want to keep the current logfile with the same name, because then i can just let my log viewer open with always the same files and have it load the current files.


EDIT: Apparently, from reading the source code, it's something completely unintuitive like this:

```json
"FileTarget": {
  "type": "File",
  "fileName": "C:\\Logfiles\\AppLog.txt",
  "archiveFileName": "C:\\Logfiles\\AppLog.txt",
  "archiveSuffixFormat": "_{1:yyyyMMdd}",
  "archiveEvery": "Day"
}
```

**Comment by snakefoot** [comment]
Yes you are right that NLog v6 includes a major rewrite of the NLog FileTarget-archive-logic, which should have its own chapter in the major-changes-list (And not just mentioned as a side-note under ConcurrentWrites-option being removed).

I had planned to wait with updating the NLog FileTarget Wiki  until NLog v6 Final version was released. But maybe I should provide a dedicate wiki-page about `archiveSuffixFormat`, which describes how to convert from old-options to new-options with NLog v6.

> @Balkoth EDIT: Apparently, from reading the source code, it's something completely unintuitive like this:

One of the goals for the NLog FileTarget rewrite was to simplify the archive-logic. Before it was possible to put archive-date / -sequence dynamically anywhere, which also made the filename-parsing / -matching logic complex. The new FileTarget has been reduced to only control the suffix being appended. To make the suffix as predictable as possible, then I decided it to be a simple string.Format without any support for NLog Layout (Hence the `{0}` and `{1}` place-holders).

The NLog team doesn't have any forum/panel where people can chip in with suggestions / ideas for how things should be built. Was also surprised how little friction/feedback there has been so far regarding the major change in a critical component like the NLog FileTarget.

I will try to provide better documentation about the major change to `archiveSuffixFormat`, and hopefully make the upgrade-path easier for everyone. But if you have ideas / suggestions for improving the NLog FileTarget and make the configuration more intuitive, then I'm most interested!

**Comment by Balkoth** [comment]
Don't take this as criticism, the library is awesome and has never let me down. I was just surprised that there is a breaking change like this and no clear path on how to upgrade. Once i figured it out my logging and even my custom targets worked without any major changes to my code.

**Comment by snakefoot** [comment]
Happy to hear that the NLog v6 upgrade went well without too much trouble. Have been testing NLog v6 with many GitHub-libraries to ensure that there would be no breaking build/runtime problems (besides obsolete warnings)

Have now updated [List of major changes in NLog v6](https://nlog-project.org/2025/04/29/nlog-6-0-major-changes.html#nlog-filetarget-and-archivesuffixformat) with a dedicated chapter about `ArchiveSuffixFormat`. I need some more time on creating a dedicated wiki-page with `ArchiveSuffixFormat`-examples. If you have any suggestions for improving the docs to make the upgrade-path easier, then please tell.

**Comment by vancodocton** [comment]
When I tried to log complex object, the serialization is not worked on publish. Any plan to support it?

Minimal repro?
```
using NLog;

LogManager.Setup()
    .LoadConfiguration(builder =>
    {
        builder.ForLogger().FilterMinLevel(LogLevel.Info).WriteToConsole();
    });

var logger = LogManager.GetCurrentClassLogger();

var complexObject = new ComplexObject()
{
    String = "string",
    Number = 1_000,
    DateTimeOffset = DateTimeOffset.UtcNow,
};

logger.Info("Log complex object: {@ComplexObject}", complexObject);

LogManager.Shutdown();

public class ComplexObject
{
    public string String { get; set; } = null!;

    public int Number { get; set; }

    public DateTimeOffset DateTimeOffset { get; set; }
}
```

**Comment by snakefoot** [comment]
@vancodocton Would it help with an extension method like this:
```c#
LogManager.Setup()
    .SetupSerialization(s => s.RegisterObjectTransformation<ComplexObject>());
    .LoadConfiguration(builder =>
    {
        builder.ForLogger().FilterMinLevel(LogLevel.Info).WriteToConsole();
    });
```
The "only" job of the `RegisterObjectTransformation` is to tell the AOT-trimming to keep public-properties for the class-type (Ex. `ComplexObject`) using `DynamicallyAccessedMemberTypes.PublicProperties`.

**Comment by vancodocton** [comment]
@snakefoot It is ok for me. Btw, I think adding new integration to `System.Text.Json` (new package, maybe) is another option, since it's already support Native AOT.

**Comment by snakefoot** [comment]
@vancodocton Yes it would be nice with better integration with `System.Text.Json`, as I expect it provides faster json-serialization without property-boxing. But I don't have the capacity to start such an adventure, but pull-requests are most wellcome.

Created #5871 to introduce new `RegisterObjectTransformation`-method that preserves public-properties. Preview-build is available here (Download to local nuget-package folder-source):

- https://ci.appveyor.com/project/nlog/nlog/builds/52197762/job/5k5kv9bfefy41axg/artifacts

**Comment by snakefoot** [comment]
NLog v6.0 RC4 has now been released, and have [compared performance](https://github.com/NLog/NLog-Performance-tests/tree/master/MicrosofLoggingPerformance) with NLog v5.5 when using Microsoft Extension Logging:

| Test Name        | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Alloc (MB) |
|------------------|-----------|-----------|-----|-----|-----|----------|------------|
| NLog v6 Async    |     3.263 | 3.064.611 |   4 | 138 | 174 |    5.937 |    3.220,2 |
| NLog v5.5 Async  |     4.084 | 2.448.374 |   7 | 166 | 222 |    7.937 |    3.815,4 |

| Test Name        | Time (ms) | Msgs/sec  | GC2 | GC1 | GC0 | CPU (ms) | Alloc (MB) |
|------------------|-----------|-----------|-----|-----|-----|----------|------------|
| NLog v6.0 Json Async  |     6.475 |   772.095 |   1 |  34 |  42 |    8.656 |    5.988,7 |
| NLog v5.5 Json Async  |     6.894 |   725.165 |   6 | 260 | 312 |   10.625 |    6.295,0 |

Grateful for those who had time to perform a test-drive, and provided their feedback. If nothing serious shows up then NLog v6 should be ready next week.

**Comment by snakefoot** [comment]
NLog v6 has now been released - https://nlog-project.org/2025/06/21/nlog-6-0-released.html

- https://www.nuget.org/packages/NLog/6.0.0#releasenotes-body-tab

**Comment by snakefoot** [comment]
Have updated the NLog Wiki with the breaking changes and new features introduced with NLog v6:

- https://github.com/NLog/NLog/pulls?q=is%3Apr+is%3Amerged+milestone%3A%226.0%22+label%3A%22needs+documentation+on+wiki%22

If one discover places where the documentation is missing an update, then please tell.

**Comment by snakefoot** [comment]
Have generated SandCastle API Docs for NLog v6:

- https://nlog-project.org/documentation/

**Comment by iratemike** [comment]
Unless I'm mistaken, doesn't `NLog.DiagnosticSource` need a package reference bump also?

**Comment by snakefoot** [comment]
> @iratemike Unless I'm mistaken, doesn't NLog.DiagnosticSource need a package reference bump also?

NLog v6 should work well with most nuget-extensions, unless they rely on obsolete NLog API that have been removed.

But yes to support AOT-builds, then NLog.DiagnosticSource probably needs to be updated.

**Comment by iratemike** [comment]
> > [@iratemike](https://github.com/iratemike) Unless I'm mistaken, doesn't NLog.DiagnosticSource need a package reference bump also?
> 
> NLog v6 should work well with most nuget-extensions, unless they rely on obsolete NLog API that have been removed.
> 
> But yes to support AOT-builds, then NLog.DiagnosticSource probably needs to be updated.

Thanks for your reply. I get an assembly load exception on startup of the application, the stack trace shows `NLog.DiagnosticSource` requesting NLog `5.0.0.0`. I've currently worked around it with a binding redirect.

**Comment by snakefoot** [comment]
> @iratemike  stack trace shows NLog.DiagnosticSource requesting NLog 5.0.0.0. I've currently worked around it with a binding redirect.

I'm guessing this is a .NET Framework issue, that is a little sensitive about strong-signed-version. I have now released [NLog.DiagnosticSource v6.0.0](https://www.nuget.org/packages/NLog.DiagnosticSource/6.0.0). Hopefully it will make your upgrade path easier.

**Comment by iratemike** [comment]
> > [@iratemike](https://github.com/iratemike)  stack trace shows NLog.DiagnosticSource requesting NLog 5.0.0.0. I've currently worked around it with a binding redirect.
> 
> I'm guessing this is a .NET Framework issue, that is a little sensitive about strong-signed-version. I have now released [NLog.DiagnosticSource v6.0.0](https://www.nuget.org/packages/NLog.DiagnosticSource/6.0.0). Hopefully it will make your upgrade path easier.

That's great, thank you!

**Comment by snakefoot** [comment]
The new NLog extensions created from the NLog v6 refactoring have been moved to their own Git-repos: #4952 

Closing this issue as the NLog v6 release is now complete. If discovering any issues with NLog v6, then please create new issue.

Thank you for all your feedback
